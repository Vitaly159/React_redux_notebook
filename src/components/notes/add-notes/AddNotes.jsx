import { useSelector, useDispatch } from "react-redux";
import { setActiveNote, onDeleteNote, setValueBody, setValueTitle } from "../../../features/addNoteSlice";
import "./addNotes.css";

function AddNotes() {
  const dispatch = useDispatch();
  const notes = useSelector((state) => state.notes.notes);
  const activeNote = useSelector((state) => state.notes.activeNote);
  const searchValue = useSelector((state) => state.notes.searchValue);

  const filteredNotes = notes.filter(
    (note) =>
      note.body.toLowerCase().includes(searchValue.toLowerCase()) ||
      note.title.toLowerCase().includes(searchValue.toLowerCase())
  );

  return filteredNotes.map((note, index) => (
    <div
      key={index}
      className={`added-note ${note.id === activeNote && "active"}`}
      onClick={() => {
        dispatch(setActiveNote(note.id));
        dispatch(setValueTitle(note.title));
        dispatch(setValueBody(note.body));
      }}
    >
      <div className="added-note-title">
        <strong>{note.title}</strong>
        <button className="del" onClick={() => dispatch(onDeleteNote(note.id))}>
          Удалить
        </button>
      </div>

      <p className="added-note-preview">
        {note.body && note.body.substr(0, 30) + "..."}
      </p>
      <small className="note-meta">
        Последнее изменение{" "}
        {new Date(note.lastChanged).toLocaleDateString("ru", {
          hour: "2-digit",
          minute: "2-digit",
        })}
      </small>
    </div>
  ));
}

export default AddNotes;
