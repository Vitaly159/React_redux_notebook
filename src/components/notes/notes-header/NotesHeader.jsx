import { useDispatch } from "react-redux";
import { onAddNote } from "../../../features/addNoteSlice";
import "./notesHeader.css";

function NotesHeader({ searching }) {
  const dispatch = useDispatch();
  
  return (
    <div>
      <div className="notes-header">
        <h1>Записи</h1>
        <button className="add-note" onClick={()=>dispatch(onAddNote())}>
          Новая запись
        </button>
      </div>
      <div>
        <input
          className="searching"
          onChange={searching}
          placeholder="Поиск по записям"
        />
      </div>
    </div>
  );
}

export default NotesHeader;
