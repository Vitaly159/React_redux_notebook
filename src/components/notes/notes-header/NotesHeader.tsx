import { useDispatch } from "react-redux";
import { onAddNote, setSearchValue } from "../../../features/addNoteSlice";
import "./notesHeader.css";

const NotesHeader: React.FC = () => {
  const dispatch = useDispatch();

  return (
    <div>
      <div className="notes-header">
        <h1>Записи</h1>
        <button className="add-note" role="save-button" onClick={() => dispatch(onAddNote(""))}>
          Новая запись
        </button>
      </div>
      <div>
        <input
          role="input"
          className="searching"
          onChange={(e) => dispatch(setSearchValue(e.target.value))}
          placeholder="Поиск по записям"
        />
      </div>
    </div>
  );
};

export default NotesHeader;
