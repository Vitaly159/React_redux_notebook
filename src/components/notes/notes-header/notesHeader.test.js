import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import NotesHeader from "./NotesHeader";
import AddNotes from "../add-notes/AddNotes";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import userEvent from "@testing-library/user-event";
import reducer, { addNoteSlice } from "../../../features/addNoteSlice";

const defaultStates = configureStore({ reducer }).getState();

const renderWithRedux = (
  component,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { notes: reducer },
      preloadedState: { notes: preloadedState },
    }),
  } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};

describe("notesHeader", () => {
  let component;
  let getByText;
  let button;
  let store;
  let input;

  beforeEach(() => {
    component = renderWithRedux(<NotesHeader />, { preloadedState: defaultStates });
    getByText = component.getByText;
    button = component.getByRole("save-button");
    input = component.getByRole("input");
    store = component.store;
  });

  test("Делаем снимок", async () => {
    expect(component.container).toMatchSnapshot();
  });
  test("Наличие кнопки добавления записи", () => {
    expect(button).toBeInTheDocument();
  });
  test("Работоспособность кнопки добавления записи", async () => {
    expect(store.getState().notes.notes.length).toEqual(0);
    await userEvent.click(button);
    const addNotesComponent = renderWithRedux(<AddNotes />, {
      preloadedState: store.getState().notes,
    });
    expect(store.getState().notes.notes.length).toEqual(1);
    expect(addNotesComponent.container).toMatchSnapshot();
  });
  test("Наличие инпута", () => {
    expect(input).toBeInTheDocument();
  });
  test("Работоспособность инпута", async () => {
    fireEvent.change(input, { target: { value: "ok ok" } });
    expect(input).toHaveValue("ok ok");
  });
});
