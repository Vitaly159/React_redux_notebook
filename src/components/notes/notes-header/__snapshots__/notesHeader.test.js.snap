// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notesHeader Делаем снимок 1`] = `
<div>
  <div>
    <div
      class="notes-header"
    >
      <h1>
        Записи
      </h1>
      <button
        class="add-note"
        role="save-button"
      >
        Новая запись
      </button>
    </div>
    <div>
      <input
        class="searching"
        placeholder="Поиск по записям"
        role="input"
      />
    </div>
  </div>
</div>
`;

exports[`notesHeader Работоспособность кнопки добавления записи 1`] = `
<div>
  <div
    class="added-note false"
    role="note-1"
  >
    <div
      class="added-note-title"
    >
      <strong>
        Имя задачи
      </strong>
      <button
        class="del"
        role="btn-rmv-1"
      >
        Удалить
      </button>
    </div>
    <p
      class="added-note-preview"
      role="start-text-1"
    >
      ...
    </p>
    <small
      class="note-meta"
    >
      Последнее изменение
       
      09.07.2023, 17:06
    </small>
  </div>
</div>
`;
